src/python/pants/engine/process.py

```diff
@@ -230,6 +235,86 @@ def remove_platform_information(res: FallibleProcessResultWithPlatform,) -> Fall
     )
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class BinaryPathRequest:
+    search_path: Tuple[str, ...]
+    binary_name: str
+
+    def __init__(self, search_path: Iterable[str], binary_name: str):
+        self.search_path = tuple(OrderedSet(search_path))
+        self.binary_name = binary_name
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class BinaryPaths(EngineAware):
+    binary_name: str
+    paths: Tuple[str, ...]
+
+    def __init__(self, binary_name: str, paths: Iterable[str]):
+        self.binary_name = binary_name
+        self.paths = tuple(OrderedSet(paths))
+
+    def level(self) -> Optional[LogLevel]:
+        return LogLevel.DEBUG if self.paths else LogLevel.WARN
+
+    def message(self) -> Optional[str]:
+        if not self.paths:
+            return f"failed to find {self.binary_name}"
+        found_msg = f"found {self.binary_name} at {self.paths[0]}"
+        if len(self.paths) > 1:
+            found_msg = f"{found_msg} and {pluralize(len(self.paths) - 1, 'other location')}"
+        return found_msg
+
+
+@rule(desc="Find binary path")
+async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
+    # TODO(John Sirois): Replace this script with a statically linked native binary so we don't
+    #  depend on either /usr/bin/env or bash being available on the Process host.
+    script_path = "./script.sh"
+    script_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    script_path,
+                    content=dedent(
+                        """
+                        #!/usr/bin/env bash
+
+                        set -euo pipefail
+
+                        if command -v which > /dev/null; then
+                            command which -a $1
+                        else
+                            command -v $1
+                        fi
+                        """
+                    ).encode(),
+                    is_executable=True,
+                )
+            ]
+        ),
+    )
+
+    paths = []
+    search_path = create_path_env_var(request.search_path)
+    result = await Get(
+        FallibleProcessResult,
+        Process(
+            description=f"Searching for `{request.binary_name}` on PATH={search_path}",
+            input_digest=script_digest,
+            argv=[script_path, request.binary_name],
+            env={"PATH": search_path},
```

Comment:
@stuhood The PATH is set here and the HermeticPex rule passes the interpreter search path the user specified in config to use for it's value.

Reply:
... the result is the full path to a python (local or remote as appropriate) to use to run the PEX with (command line arg0). This all has 0 to do with setting the PATH of the PEX process. That is simply set via config option.

Reply:
> That is simply set via config option.

Right: which is what I was saying about needing to hardcode (a misnomer I guess: "configure') the PATH of the remote host.

Reply:
But yea, I see now, thanks. The path hardcoding is an independent issue that we can tackle via #9760, and the "binary discovery" within that PATH should be portable. Will respond in the other thread.